packages/core/src/api/middleware/auth-guard.ts

```diff
@@ -35,9 +40,31 @@ export class AuthGuard implements CanActivate {
         const isPublic = !!permissions && permissions.includes(Permission.Public);
         const hasOwnerPermission = !!permissions && permissions.includes(Permission.Owner);
         const session = await this.getSession(req, res, hasOwnerPermission);
-        const requestContext = await this.requestContextService.fromRequest(req, info, permissions, session);
+        let requestContext = await this.requestContextService.fromRequest(req, info, permissions, session);
         (req as any)[REQUEST_CONTEXT_KEY] = requestContext;
 
+        // In case the session does not have an activeChannelId or the activeChannelId
+        // does not correspond to the current channel, the activeChannelId on the session is set
+        if (session && (!session.activeChannelId || session.activeChannelId !== requestContext.channelId)) {
+            await this.sessionService.setActiveChannel(session, requestContext.channel);
+            if (requestContext.activeUserId) {
+                const customer = await this.customerService.findOneByUserId(
+                    requestContext,
+                    requestContext.activeUserId,
+                    false,
+                );
+                // To avoid assigning the customer to the active channel on every request,
+                // it is only done on the first request and whenever the channel changes
+                if (customer) {
+                    await this.channelService.assignToChannels(Customer, customer.id, [
+                        requestContext.channelId,
+                    ]);
+                }
+            }
+            requestContext = await this.requestContextService.fromRequest(req, info, permissions, session);
```

Comment:
Ah much better than the recursion 👍.
I would suggest extracting the first part of this `if` block (ln 49-63) into a well-named method to make it easier to follow the high-level logic of the outer `canActivate` method. E.g.:

```ts
private setActiveChannel(session: CachedSession, requestContext: RequestContext): CachedSession {
  // ...
  }
```

and also think about extracting the `if` expression into a named constant like `activeChannelShouldBeSet` for better readability.